src/libcore/slice/mod.rs

```diff
@@ -3021,14 +3021,20 @@ macro_rules! iterator {
                 unsafe { from_raw_parts(self.ptr, len!(self)) }
             }
 
+            // Helper function for shrinking the iterator when T is a ZST, by moving the
+            // end of the iterator backwards by `n`.
+            #[inline(always)]
+            fn zst_shrink(&mut self, n: isize) {
+                self.end = (self.end as * $raw_mut u8).wrapping_offset(-n) as * $raw_mut T;
```

Comment:
Maybe add a `debug_assert` making sure `mem::size_of::<T>() == 0`?

Reply:
I didn't do this as not to affect the performance of debug builds, although I'm not opposed to adding it. Either way, this function should probably be `unsafe`?